sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -866,71 +867,233 @@ class Analyzer(
+     * attributes.
+     */
+    private def rewriteSubQuery(
+        sub: LogicalPlan,
+        outer: LogicalPlan): (LogicalPlan, Seq[Expression]) = {
+      // Pull out the tagged predicates and rewrite the subquery in the process.
+      val (basePlan, baseConditions) = pullOutCorrelatedPredicates(sub)
+
+      // Make sure the inner and the outer query attributes do not collide.
+      val duplicates = basePlan.outputSet.intersect(outer.outputSet)
+      val (plan, deDuplicatedConditions) = if (duplicates.nonEmpty) {
+        val aliasMap = AttributeMap(duplicates.map { dup =>
+          dup -> Alias(dup, dup.toString)()
+        }.toSeq)
+        val aliasedExpressions = basePlan.output.map { ref =>
+          aliasMap.getOrElse(ref, ref)
+        }
+        val aliasedProjection = Project(aliasedExpressions, basePlan)
+        val aliasedConditions = baseConditions.map(_.transform {
+          case ref: Attribute => aliasMap.getOrElse(ref, ref).toAttribute
+        })
+        (aliasedProjection, aliasedConditions)
+      } else {
+        (basePlan, baseConditions)
+      }
+      // Remove outer references from the correlated predicates. We wait with extracting
+      // these until collisions between the inner and outer query attributes have been
+      // solved.
+      val conditions = deDuplicatedConditions.map(_.transform {
+        case OuterReference(ref) => ref
+      })
+      (plan, conditions)
+    }
+
+    /**
+     * Rewrite a LogicalPlan containing [[ScalarSubquery]], [[ListQuery]] or [[Exists]] expressions.
+     */
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case q: UnaryNode if q.childrenResolved =>
+        q transformExpressions {
+          case ScalarSubquery(sub, None, exprId) if sub.resolved && sub.output.nonEmpty =>
+            val (rewrite, conditions) = rewriteSubQuery(sub, q.child)
+            ScalarSubquery(rewrite, Some(conditions), exprId)
+          case Exists(sub, exprId) if sub.resolved =>
+            val (rewrite, conditions) = rewriteSubQuery(sub, q.child)
+            PredicateSubquery(rewrite, conditions, nullAware = false, exprId)
+          case In(e, Seq(ListQuery(sub, exprId))) if sub.resolved && e.resolved =>
+            val (rewrite, conditions) = rewriteSubQuery(sub, q.child)
+            // Get the left hand side expressions.
+            val expressions = e match {
+              case CreateStruct(exprs) => exprs
+              case expr => Seq(expr)
+            }
+            // Make sure the number of arguments are equal.
+            if (expressions.size != sub.output.size) {
+              failAnalysis(s"The number of fields in the value (${expressions.size}) does not " +
+                s"match with the number of columns in the subquery (${sub.output.size})")
+            }
+            // Construct the IN conditions.
+            val inConditions = expressions.zip(rewrite.output).map(EqualTo.tupled)
```

Comment:
nwm, we had make sure that the two sides of plan do not have duplicated attributes